streamer/src/nonblocking/quic.rs

```diff
@@ -825,25 +825,37 @@ async fn handle_connection(
                             params.total_stake,
                         );
 
-                    stream_counter.reset_throttling_params_if_needed();
-                    if stream_counter.stream_count.load(Ordering::Relaxed)
-                        >= max_streams_per_throttling_interval
-                    {
-                        stats.throttled_streams.fetch_add(1, Ordering::Relaxed);
-                        match params.peer_type {
-                            ConnectionPeerType::Unstaked => {
-                                stats
-                                    .throttled_unstaked_streams
-                                    .fetch_add(1, Ordering::Relaxed);
-                            }
-                            ConnectionPeerType::Staked(_) => {
-                                stats
-                                    .throttled_staked_streams
-                                    .fetch_add(1, Ordering::Relaxed);
+                    let throttle_interval_start =
+                        stream_counter.reset_throttling_params_if_needed();
+                    let streams_read_in_throttle_interval =
+                        stream_counter.stream_count.load(Ordering::Relaxed);
+                    if streams_read_in_throttle_interval >= max_streams_per_throttling_interval {
+                        // The peer is sending faster than we're willing to read. Sleep for what's
+                        // left of this read interval so the peer backs off.
+                        let throttle_duration = STREAM_THROTTLING_INTERVAL
+                            .saturating_sub(throttle_interval_start.elapsed())
+                            .saturating_sub(connection.rtt());
```

Comment:
Can we explain in a comment why we are subtracting `connectio.rtt()`? It will help with future maintainability of the code.

Reply:
The comment could say something like:

```
// ... so the peer backs off.
//
// We subtract `connection.rtt()` to the time we sleep to minimize drift. That's the estimated amount
// of time needed for us to send credits back to the peer, and for the peer to send the next chunk of
// data so that it falls near the start of the next interval window.
```

Reply:
If that's the reason, we should reduce the throttle duration just before calling sleep. The current code is actually increasing the total throttling window by the rtt time

Reply:
What's average value for rtt. Wondering if this will reduce the throttling window enough that throttling never happens.

Reply:
It's per connection and the quic protocol describes how to compute and update it. It's meant to be a good estimate of the actual rtt and is used internally by quinn to piggyback control frames to minimize overhead. It can only be way off if there's a bug in quinn, but at that point, many other things wouldn't work as intended anyway.

Reply:
We want to read X streams each interval of T ms. We just found that we've read X streams and we're at time T1<T.

Ideally, we wanted to have received this stream X at time T, not T1. So we're going to sleep to slow down the peer, and we want them to send the next stream in the *next* interval window of T ms. The difference between the time we received the stream and the time we wanted to receive it is T-T1. That's STREAM_THROTTLING_INTERVAL -  throttle_interval_start.elapsed().

If we sleep for T - T1, *we* will send a frame back to the peer at time T which tells them that they can start sending again. That will take rtt / 2 time to reach the peer, which will start sending again, so we'll receive the next stream around T + rtt. We want to receive the next stream *approximately* at time T tho (the beginning of the next interval), therefore we also subtract rtt.

Where is the logic wrong?

Reply:
For a well behaving client it would be ok. I am wondering if a malicious client can implement/hack a custom quic client that causes high rtt but also has high stream count. A simple mitigation would be to cap the rtt value used in this calculation.

Reply:
The consideration of RTT makes sense only when you assume the streams are sent sequentially. It is not true, the client can send the streams simultaneously. While we think it will take a RTT to get the next stream, the next stream might have already arrived. I think the logic is simpler just consider this stream which is causing the violation -- not do the optimization for the next stream which we do not know when it might come.

Reply:
> For a well behaving client it would be ok. I am wondering if a malicious client can implement/hack a custom quic client that causes high rtt but also has high stream count. A simple mitigation would be to cap the rtt value used in this calculation.

I'll check the spec, but yeah you're right in principle a malicious client could intentionally game the rtt to make his connection more bursty. Not sure what a good cap for rtt could be.

Reply:
@lijunwangs 
>  While we think it will take a RTT to get the next stream, the next stream might have already arrived

No, because by sleeping if the peer keeps sending the receive window will fill up, which is the whole point of sleeping. So the next stream _the peer sends_ will be after our sleeping time. If in the meantime there are queued streams within the receive window that's fine.

Reply:
> Not sure what a good cap for rtt could be.

I say for the purpose of landing this soon, let's remove rtt as a variable. Then we can play with taking rtt in consideration for computing the receive window and/or the throttle time and we can let that ride the trains.

Reply:
done